File: DFS/#530 Minimum Absolute Difference in BST.cpp
//Track minimum as proceeding inorder-traversal (faster)
class Solution {
    public:
    int minDiff = INT_MAX;
    void inorder(TreeNode* root, int &val){
        if(!root) return;
        inorder(root->left,val);
        minDiff=min(minDiff,abs(root->val - val));
        val = root->val;
        inorder(root->right,val);
    }
    

    int getMinimumDifference(TreeNode* root) {
        int val = INT_MAX;
        inorder(root,val);
        return minDiff;
    }
};

//Another way is to store all the values in the vector
class Solution {
    public:
    vector<int> res;
    void inorder(TreeNode* root){
        if(!root) return;
        inorder(root->left);
        res.push_back(root->val);
        inorder(root->right);
    }
    int minDiff(){
        int minVal=INT_MAX;
        for(int i=0; i<res.size()-1; i++){
            minVal = min(minVal,res[i+1]-res[i]);
        }
        return minVal;
    }

    int getMinimumDifference(TreeNode* root) {
        inorder(root);
        return minDiff();
    }
};

File: Tree/#530. Minimum Absolute Difference in BST.py
class Solution:
    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        self.prev = None
        self.min_diff = math.inf
        self.inorder(root)
        return self.min_diff

    def inorder(self, root):
        if not root:
            return None
        self.inorder(root.left)
        if self.prev is not None:
            self.min_diff = min(self.min_diff, root.val - self.prev)
        self.prev = root.val
        self.inorder(root.right)
        